src/api.rs

```diff
@@ -291,47 +295,47 @@ impl Config {
   }
 
   pub fn new_context(&self) -> Context {
-    let seq = Sequence::new(&self.frame_info);
-    let fi = FrameInvariants::new(
-      self.frame_info.width,
-      self.frame_info.height,
-      self.enc,
-      seq,
-    );
-
     #[cfg(feature = "aom")]
     unsafe {
       av1_rtcd();
       aom_dsp_rtcd();
     }
 
     Context {
-      fi,
       frame_count: 0,
       frames_to_be_coded: 0,
       idx: 0,
+      frames_processed: 0,
       frame_q: BTreeMap::new(),
+      frame_data: BTreeMap::new(),
+      keyframes: BTreeSet::new(),
       packet_data: Vec::new(),
       segment_start_idx: 0,
       segment_start_frame: 0,
-      frame_types: BTreeMap::new(),
       keyframe_detector: SceneChangeDetector::new(&self.frame_info),
+      config: *self,
     }
   }
 }
 
 pub struct Context {
-  fi: FrameInvariants,
   //    timebase: Rational,
   frame_count: u64,
   frames_to_be_coded: u64,
   idx: u64,
+  frames_processed: u64,
+  /// Maps frame *number* to frames
   frame_q: BTreeMap<u64, Option<Arc<Frame>>>, //    packet_q: VecDeque<Packet>
+  /// Maps frame *idx* to frame data
+  frame_data: BTreeMap<u64, FrameInvariants>,
```

Comment:
Thanks for converting FrameInvariants to an array. Would it make sense to convert both frame_q and frame_data to both use either number or idx?

Reply:
I had hoped to accomplish that, however I struggled to get that to work due to how frame reordering works--the fact that we don't know which idx links to which frame number until we've computed the invariants for that frame.